File: C++/Binary Tree/Binary tree.cpp
#include<iostream>
using namespace std;

typedef int element;

typedef struct Node
{
    element data;
    struct Node *left, *right;
}Node;

Node* makeNode(element item)
{
    Node* newNode = (Node *)malloc(sizeof(Node));
    newNode->data = item;
    
    return newNode;
}

void fillTree(Node* root, Node* l, Node* r)
{
    root->left = l;
    root->right = r;
}
                     
void preorder(Node *root)
{
    if(root != NULL)
    {
        cout << "("<<root->data<<") ";
        preorder(root->left);
        preorder(root->right);
    }
}

void inorder(Node *root)
{
    if(root != NULL)
    {
        inorder(root->left);
        cout << "("<<root->data<<") ";
        inorder(root->right);
    }
}

void postorder(Node *root)
{
    if(root != NULL)
    {
        postorder(root->left);
        postorder(root->right);
        cout << "("<<root->data<<") ";
    }
}

int main()
{
   Node *node[7], *root;
    
   for(int i=0; i<7; ++i) {
      node[i] = makeNode(i+1);
   }

   root = node[0];

    fillTree(root, node[1], node[2]);
    fillTree(node[1], node[3], nullptr);
    fillTree(node[2], node[4], node[5]);
    fillTree(node[3], NULL, NULL);
    fillTree(node[4], NULL, node[6]);
    fillTree(node[5], NULL, NULL);
    fillTree(node[6], NULL, NULL);
    
    cout << "===preorder===\n";
    preorder(root);
    cout << "\n===inorder===\n";
    inorder(root);
    cout << "\n===postorder===\n";
    postorder(root);
}

File: Python/Tree/Binary Tree/Binary Tree.py
class MakeNode:
    def __init__(self, item, left=None, right=None):
        self.data = item
        self.left = left
        self.right = right


class Node:
    def __init__(self, item):
        self.root = MakeNode(item)

    def insert_left(self, node):
        self.root.left = node

    def insert_right(self, node):
        self.root.right = node


def preorder(node):
    if node is not None:
        print("({})". format(node.root.data), end=" ")
        preorder(node.root.left)
        preorder(node.root.right)


def inorder(node):
    if node is not None:
        inorder(node.root.left)
        print("({})".format(node.root.data), end=" ")
        inorder(node.root.right)


def postorder(node):
    if node is not None:
        postorder(node.root.left)
        postorder(node.root.right)
        print("({})".format(node.root.data), end=" ")


if __name__ == '__main__':

    node = [None for _ in range(7)]

    for idx in range(1, 7):
        node[idx] = Node(idx)

    # link two nodes
    node[1].insert_left(node[2])
    node[1].insert_right(node[3])

    node[2].insert_left(node[4])
    node[2].insert_right(node[5])

    node[3].insert_left(node[6])

    print("===preorder===")
    preorder(node[1])
    print("\n")

    print("===inorder===")
    inorder(node[1])
    print("\n")

    print("===postorder")
    postorder(node[1])










